src/flint/types/nmod_mat.pyx

```diff
@@ -530,8 +571,8 @@ cdef class nmod_mat(flint_mat):
         if self.nrows() != self.ncols():
             raise ValueError("fmpz_mod_mat charpoly: matrix must be square")
 
-        res = nmod_poly.__new__(nmod_poly)
-        nmod_poly_init(res.val, self.val.mod.n)
+        # XXX: don't create a new context for the polynomial
+        res = nmod_poly_new_init(any_as_nmod_poly_ctx(self.ctx.mod.n))
         nmod_mat_charpoly(res.val, self.val)
```

Comment:
Need to avoid creating new contexts. I think we need an `nmod_mat_ctx` that holds an `nmod_ctx` and an `nmod_poly_ctx` as attributes.

Reply:
Now an `nmod_mat_ctx` holds an `nmod_poly_ctx` which holds an `nmod_ctx`.